doh-server/server.go

```diff
@@ -107,12 +111,46 @@ func (s *Server) Start() error {
 	if s.conf.Verbose {
 		servemux = handlers.CombinedLoggingHandler(os.Stdout, servemux)
 	}
+
+	var clientCAPool *x509.CertPool
+	if s.conf.TLSClientAuth && s.conf.CertCA != "" {
+		clientCA, err := ioutil.ReadFile(s.conf.CertCA)
+		if err != nil {
+			log.Fatalf("reading cert failed : %v", err)
+		}
+		clientCAPool = x509.NewCertPool()
+		clientCAPool.AppendCertsFromPEM(clientCA)
+		log.Println("Client cert loaded for TLS authentication")
+	}
+
 	results := make(chan error, len(s.conf.Listen))
 	for _, addr := range s.conf.Listen {
 		go func(addr string) {
 			var err error
+			log.Println("start server")
+
 			if s.conf.Cert != "" || s.conf.Key != "" {
-				err = http.ListenAndServeTLS(addr, s.conf.Cert, s.conf.Key, servemux)
+				if s.conf.TLSClientAuth && s.conf.CertCA != "" {
```

Comment:
Why `if s.conf.TLSClientAuth && s.conf.CertCA != ""`?
What the point to use `&&` here?

Reply:
we need the certca to perform the authentication of the client

Reply:
> we need the certca to perform the authentication of the client

So if TLSClientAuth is true, but CertCA is "", what will happen?
You should not put `&&` here.

Reply:
with TLSClientAuth is true, but CertCA is "", authentication should happen

Reply:
> with TLSClientAuth is true, but CertCA is "", authentication should not happen

But you don't print out an error message, which may become a security problem.